pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -156,12 +156,20 @@ func (t *ResolvedPipelineRunTask) checkParentsDone(facts *PipelineRunFacts) bool
 }
 
 func (t *ResolvedPipelineRunTask) skip(facts *PipelineRunFacts) bool {
-	if facts.isFinalTask(t.PipelineTask.Name) || t.IsStarted() {
+	if t.IsStarted() {
 		return false
 	}
 
-	if t.conditionsSkip() || t.whenExpressionsSkip(facts) || t.parentTasksSkip(facts) || facts.IsStopping() {
-		return true
+	if facts.isFinalTask(t.PipelineTask.Name) {
+		if t.whenExpressionsSkip(facts) {
+			return true
+		}
+	}
+
+	if facts.isDAGTask(t.PipelineTask.Name) {
+		if t.conditionsSkip() || t.whenExpressionsSkip(facts) || t.parentTasksSkip(facts) || facts.IsStopping() {
+			return true
+		}
```

Comment:
nit: is it possible to word these as a switch statement? feels weird that facts.isFinalTask can be true and yet facts.isDAGTask(t.PipelineTask.Name) still gets evaluated

Reply:
simplified the logic and removed the double checks

Reply:
This could be simplified by utilizing `IsFinallySkipped` instead of changing this function.

https://github.com/tektoncd/pipeline/blob/01286cf978750e85609fa0a0e9a1aa177fc724f8/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L220

This skip function is primarily used to build cache of skipped `dag` tasks. This caching was introduced in PR -https://github.com/tektoncd/pipeline/pull/3524

https://github.com/tektoncd/pipeline/blob/01286cf978750e85609fa0a0e9a1aa177fc724f8/pkg/reconciler/pipelinerun/resources/pipelinerunstate.go#L59

Reply:
Just introduce one `if` statement after [line 227](https://github.com/tektoncd/pipeline/blob/01286cf978750e85609fa0a0e9a1aa177fc724f8/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L227) 

```
		if t.whenExpressionsSkip(facts) {
			return true
		}

```